### DISCUSSION TITLE
Access compiler functions outside of compiler

### DISCUSSION THREAD

**Discussion by tomhillgreyridge** [unpoly/unpoly#398] [Q&A] [answered]
Just wondering whether there was an "idiomatic" way of doing something.  

Suppose you have an up-compiler which is used to implement a calendar (just an example)

```javascript
up.compiler('div.calendar',(element) => {
  function show(){
    //logic for showing calendar
  }
  function hide(){
    //logic for hiding calendar
  }
  //bind click handlers for opening and closing calendar here
);
```
This is great and gives me a nice encapsulated set of behaviour.  However, if I now want to trigger the opening of the calendar from somewhere else (i.e. programatically) is there a good way of doing it?

- If I know the event handlers I could trigger a click on the relevant button which is used for opening the calendar
- I could make the compiler listen to a custom event but that is complex if there is more than one calendar

My current thought is to basically create the compiler as a class and store it against the element

```javascript
class MyCalendar{
  constructor(element){
    this.element = element;

    //add event handlers here
 }
 show(){
   //stuff for showing the calendar here
 }
 hide(){
  //stuff for hiding the calendar here
 }
}

up.compiler('div.calendar',(element) => { 
  element.compiler = new MyCalendar(element);
});
```

That way, if I have the element I can simply ask for its compiler and call whatever functions on it I need to.

**Comment by triskweline** [comment, marked as answer]
I often do something similar where I only expose the public API as methods on the element itself:

```js
up.compiler('div.calendar',(element) => { 
  function show() {
    //logic for showing calendar
  }
  function hide() {
    //logic for hiding calendar
  }

  element.show = show
  element.hide = hide
})
```

**Reply by tomhillgreyridge** [reply]
Okay, great.  I like your solution of exposing just the public API.  I might put them inside an "api" namespace on the element just so I don't pollute the element with too many functions i.e. element.api.show() and element.api.hide()

One other thing, am I right in thinking that if we add / remove elements to the DOM via javascript we need to manually call up.hello() or up.destroy() on them in order for compilers to be correctly hooked up?  I wasn't sure whether Unpoly was using mutation observers to do this automatically or not.
